compiler/src/iree/compiler/GlobalOptimization/FuseDequantizationMatmul.cpp

```diff
@@ -0,0 +1,860 @@
+//                                                d1, d2) -> (d0, d1)>],
+//                               iterator_types = ["parallel", "parallel",
+//                               "reduction"]} ins(%17, %0 : tensor<4x2xi16>,
+//                               tensor<8x4x2xi4>) outs(%19 : tensor<8x4xi32>) {
+//             ^bb0(%in: i16, %in_4: i4, %out: i32):
+//               %24 = arith.extsi %in : i16 to i32
+//               %25 = arith.extui %in_4 : i4 to i32
+//               %26 = arith.muli %24, %25 : i32
+//               %27 = arith.addi %26, %out : i32
+//               linalg.yield %27 : i32
+//             } -> tensor<8x4xi32>
+//         ```
+//         This op also extends the inputs to the accumulation type, i32 in this
+//         case, to target specific x86 instructions. We perform the matrix
+//         multiplication before the dequantization arithmetic, which has been
+//         reassociated into op 6.
+//      6. The final op performs the remaining reduction across groups and does
+//      the
+//         dequantization arithmetic:
+//         ```mlir
+//         %23 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0,
+//         d1)>,
+//                                                affine_map<(d0, d1) -> (d1)>,
+//                                                affine_map<(d0, d1) -> (d1)>,
+//                                                affine_map<(d0, d1) -> (d0,
+//                                                d1)>, affine_map<(d0, d1) ->
+//                                                (d0, d1)>, affine_map<(d0, d1)
+//                                                -> (d0)>],
+//                               iterator_types = ["parallel", "reduction"]}
+//                               ins(tensor<8x4xi32>, tensor<4xf32>,
+//                               tensor<4xf32>, tensor<8x4xf32>,
+//                               tensor<8x4xf32>) outs(tensor<8xf32>) {
+//             ^bb0(%in: i32, %in_4: f32, %in_5: f32, %in_6: f32, %in_7: f32,
+//             %out: f32):
+//               %24 = arith.sitofp %in : i32 to f32
+//               %25 = arith.mulf %24, %in_4 : f32
+//               %26 = arith.mulf %25, %in_6 : f32
+//               %27 = arith.mulf %in_7, %in_6 : f32
+//               %28 = arith.mulf %27, %in_5 : f32
+//               %29 = arith.subf %26, %28 : f32
+//               %30 = arith.addf %29, %out : f32
+//               linalg.yield %30 : f32
+//             } -> tensor<8xf32>
+//         ```
+//
+// The rewrite also forms a `flow.dispatch.region` op around ops 5 and 6, and
+// sets a specific tiling on ops 5 and 6 to target a VectorContractCustomKernel.
+//
+// ** Note that this rewrite introduces precision loss in the matmul, and is a
+//    tradeoff between precision and performance. This rewrite should most
+//    likely be opt-in only. **
+static LogicalResult reassociateAndFuseDequantMatmul(RewriterBase &rewriter,
```

Comment:
This method is too long. Maybe worth splitting out as the steps above suggest to make the overal logic more readable.

Reply:
Sure, I'll break this down. Will make it easier to maintain this later too